diff --git a/.rh_rpm_package b/.rh_rpm_package
index c08b6ad..672f66a 100644
--- a/.rh_rpm_package
+++ b/.rh_rpm_package
@@ -1 +1 @@
-7.2.8++
+7.2.9
diff --git a/README b/README
index bfbaef6..e2af924 100644
--- a/README
+++ b/README
@@ -33,7 +33,6 @@
   can be found here:
  
          https://crash-utility.github.io/crash_whitepaper.html
-         
  
   These are the current prerequisites: 
 
@@ -75,8 +74,8 @@
 
   To build the crash utility: 
 
-    $ tar -xf crash-7.2.8.tar.gz
-    $ cd crash-7.2.8
+    $ tar -xf crash-7.2.9.tar.gz
+    $ cd crash-7.2.9
     $ make
 
   The initial build will take several minutes  because the embedded gdb module
@@ -119,7 +118,7 @@
   If neither /dev/mem or /dev/crash are available, then /proc/kcore will be
   be used as the live memory source.  If /proc/kcore is also restricted, then
   the Red Hat /dev/crash driver may be compiled and installed; its source
-  is included in the crash-7.2.8/memory_driver subdirectory.
+  is included in the crash-7.2.9/memory_driver subdirectory.
 
   If the kernel file is stored in /boot, /, /boot/efi, or in any /usr/src
   or /usr/lib/debug/lib/modules subdirectory, then no command line arguments
@@ -130,7 +129,7 @@
 
     $ crash
     
-    crash 7.2.8
+    crash 7.2.9
     Copyright (C) 2002-2020  Red Hat, Inc.
     Copyright (C) 2004, 2005, 2006, 2010  IBM Corporation
     Copyright (C) 1999-2006  Hewlett-Packard Co
@@ -155,7 +154,7 @@
           KERNEL: /boot/vmlinux
         DUMPFILE: /dev/mem
             CPUS: 1
-            DATE: Thu Jan 30 13:58:56 2020
+            DATE: Fri Nov 20 08:31:51 JST 2020
           UPTIME: 10 days, 22:55:18
     LOAD AVERAGE: 0.08, 0.03, 0.01
            TASKS: 42
@@ -183,7 +182,7 @@
     eval           kmem           ptob           sym            wr             
     exit           list           ptov           sys            q              
     
-    crash version: 7.2.8    gdb version: 7.6
+    crash version: 7.2.9    gdb version: 7.6
     For help on any command above, enter "help <command>".
     For help on input options, enter "help input".
     For help on output options, enter "help output".
@@ -196,7 +195,7 @@
 
     $ crash vmlinux vmcore
  
-    crash 7.2.8
+    crash 7.2.9
     Copyright (C) 2002-2020  Red Hat, Inc.
     Copyright (C) 2004, 2005, 2006, 2010  IBM Corporation
     Copyright (C) 1999-2006  Hewlett-Packard Co
